File: lingting-core/src/main/java/live/lingting/framework/thread/AbstractThreadContextComponent.java
package live.lingting.framework.thread;

import live.lingting.framework.context.ContextComponent;
import live.lingting.framework.context.ContextHolder;
import live.lingting.framework.util.StringUtils;
import live.lingting.framework.util.ThreadUtils;
import live.lingting.framework.util.ValueUtils;
import live.lingting.framework.value.WaitValue;
import lombok.SneakyThrows;
import org.slf4j.Logger;

import java.util.concurrent.Executor;
import java.util.function.Consumer;
import java.util.function.Function;

/**
 * @author lingting 2023-04-22 10:40
 */
@SuppressWarnings("java:S2142")
public abstract class AbstractThreadContextComponent implements ContextComponent {

	protected final Logger log = org.slf4j.LoggerFactory.getLogger(getClass());

	protected final WaitValue<Thread> threadValue = WaitValue.of();

	protected void thread(Consumer<Thread> consumer) {
		thread(thread -> {
			consumer.accept(thread);
			return null;
		}, null);
	}

	protected <T> T thread(Function<Thread, T> function, T defaultValue) {
		if (!threadValue.isNull()) {
			Thread value = threadValue.getValue();
			return function.apply(value);
		}
		return defaultValue;
	}

	protected long threadId() {
		return thread(Thread::getId, -1L);
	}

	@SneakyThrows
	protected void interrupt() {
		thread(t -> {
			if (!t.isInterrupted()) {
				t.interrupt();
			}
		});
		threadValue.update((Thread) null);
	}

	protected void init() {

	}

	public boolean isRun() {
		boolean threadAvailable = thread(thread -> !thread.isInterrupted() && thread.isAlive(), false);
		return threadAvailable && !ContextHolder.isStop();
	}

	protected Executor executor() {
		return ThreadUtils.executor();
	}

	@Override
	public void onApplicationStart() {
		String name = getSimpleName();
		Executor executor = executor();
		executor.execute(new KeepRunnable(name) {
			@Override
			protected void process() throws Throwable {
				Thread thread = Thread.currentThread();
				threadValue.update(thread);
				try {
					AbstractThreadContextComponent.this.run();
				}
				finally {
					threadValue.update((Thread) null);
				}
			}
		});
	}

	@Override
	public void onApplicationStop() {
		log.warn("Class: {}; ThreadId: {}; closing!", getSimpleName(), threadId());
		interrupt();
	}

	public String getSimpleName() {
		String simpleName = getClass().getSimpleName();
		if (StringUtils.hasText(simpleName)) {
			return simpleName;
		}
		return getClass().getName();
	}

	public void run() {
		init();
		while (isRun()) {
			try {
				doRun();
			}
			catch (InterruptedException e) {
				interrupt();
				shutdown();
			}
			catch (Exception e) {
				error(e);
			}
		}
	}

	@SuppressWarnings("java:S112")
	protected abstract void doRun() throws Exception;

	/**
	 * 线程被中断触发.
	 */
	protected void shutdown() {
		log.warn("Class: {}; ThreadId: {}; shutdown!", getSimpleName(), threadId());
	}

	protected void error(Exception e) {
		log.error("Class: {}; ThreadId: {}; error!", getSimpleName(), threadId(), e);
	}

	public void awaitTerminated() {
		log.debug("wait thread terminated.");
		ValueUtils.awaitTrue(() -> thread(thread -> Thread.State.TERMINATED.equals(thread.getState()), true));
		log.debug("thread terminated.");
	}

}

File: lingting-core/src/main/kotlin/live/lingting/framework/thread/AbstractThreadContextComponent.kt
package live.lingting.framework.thread

import java.util.concurrent.Executor
import java.util.concurrent.ExecutorService
import java.util.function.Consumer
import live.lingting.framework.context.ContextComponent
import live.lingting.framework.context.ContextHolder
import live.lingting.framework.kt.ifTrue
import live.lingting.framework.kt.logger
import live.lingting.framework.util.StringUtils
import live.lingting.framework.util.ThreadUtils
import live.lingting.framework.util.ValueUtils
import live.lingting.framework.value.WaitValue

/**
 * @author lingting 2023-04-22 10:40
 */
abstract class AbstractThreadContextComponent : ContextComponent {
    protected val log = logger()

    val threadValue: WaitValue<Thread> = WaitValue.of<Thread>()

    private var executor: ExecutorService = VirtualThread.executor()

    protected fun thread(consumer: Consumer<Thread>) {
        threadValue.optional().ifPresent { consumer.accept(it) }
    }

    protected fun threadId(): Long {
        return threadValue.optional().map { it.threadId() }.orElse(-1L)
    }

    protected fun interrupt() {
        threadValue.consumer {
            it?.isInterrupted.ifTrue { it?.interrupt() }
        }
    }

    protected fun init() {
    }

    open val isRun: Boolean
        get() {
            val available = threadValue.optional().map { !it.isInterrupted && it.isAlive }.orElse(false)
            return !ContextHolder.isStop && available
        }

    protected open fun executor(): Executor {
        return executor
    }

    fun useThreadPool() {
        executor = ThreadUtils.executor()
    }

    fun useThreadVirtual() {
        executor = VirtualThread.executor()
    }

    override fun onApplicationStart() {
        val name = simpleName
        threadValue.consumer {
            // 已分配, 不再重新分配
            if (it != null) {
                return@consumer
            }

            executor().execute(object : KeepRunnable(name) {
                override fun process() {
                    val thread = Thread.currentThread()
                    threadValue.update(thread)
                    try {
                        this@AbstractThreadContextComponent.run()
                    } finally {
                        threadValue.update(null)
                    }
                }
            })
        }
    }

    override fun onApplicationStop() {
        log.warn("Class: {}; ThreadId: {}; closing!", simpleName, threadId())
        interrupt()
    }

    val simpleName: String
        get() {
            val simpleName: String = javaClass.getSimpleName()
            if (StringUtils.hasText(simpleName)) {
                return simpleName
            }
            return javaClass.getName()
        }

    fun run() {
        init()
        while (isRun) {
            try {
                doRun()
            } catch (_: InterruptedException) {
                interrupt()
                shutdown()
            } catch (e: Exception) {
                error(e)
            }
        }
    }


    protected abstract fun doRun()

    /**
     * 线程被中断触发.
     */
    protected open fun shutdown() {
        log.warn("Class: {}; ThreadId: {}; shutdown!", simpleName, threadId())
    }

    protected fun error(e: Exception) {
        log.error("Class: {}; ThreadId: {}; error!", simpleName, threadId(), e)
    }

    fun awaitTerminated() {
        log.debug("wait thread terminated.")
        ValueUtils.awaitTrue { threadValue.optional().map { Thread.State.TERMINATED == it.state }.orElse(true) }
        log.debug("thread terminated.")
    }
}
